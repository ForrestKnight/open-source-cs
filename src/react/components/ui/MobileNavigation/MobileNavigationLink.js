/* Packages */
import React from 'react';
import {NavLink} from 'react-router-dom';
import {connect} from 'react-redux';


/* Actions */
import {navMenu_Mobile__Toggle} from '../../../redux/actions/ui';


/* Component */
export class MobileNavigationLink extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            sublink_Menu: false
        };
    };


    sublink_Menu__Toggle = () => {
        this.setState({
            sublink_Menu: !this.state.sublink_Menu
        });
    };


    navMenu_Mobile__Toggle = () => {
        /* Disable scrolling */
        const html = document.getElementsByTagName('HTML')[0];
        html.setAttribute('data-Mobile_Nav', 'false');

        /* Dispatch */
        this.props.navMenu_Mobile__Toggle();
    };


    render() {
        return (
            <li className="MobileNavigationLink">
                {this.props.internal == true ?
                    /* Internal link */
                    <NavLink 
                        className="MobileNavigationLink__link"
                        to={this.props.href}
                        onClick={this.navMenu_Mobile__Toggle}
                    >
                        {this.props.text}
                    </NavLink>
                    :

                    /* External link */
                    <a
                        className="MobileNavigationLink__link"
                        href={this.props.href}
                        onClick={this.navMenu_Mobile__Toggle}
                        target="_blank"
                        rel="noopener"
                    >
                        {this.props.text}
                    </a>
                }

                {/* Sublink button */}
                {this.props.sublinks &&
                    <button
                        className="MobileNavigationLink__sublink_button"
                        onClick={this.sublink_Menu__Toggle}
                        data-active={this.state.sublink_Menu}
                    >
                        â–¼
                    </button>
                }

                {/* Sublink menu */}
                {this.props.sublinks &&
                    <ul
                        className="MobileNavigationLink__sublink_list"
                        onClick={this.sublink_Menu__Toggle}
                        data-active={this.state.sublink_Menu}
                    >
                        {this.props.sublinks.map((sublink) => (
                            <li
                                className="MobileNavigationLink__sublink"
                                key={sublink.text}
                            >
                                {sublink.internal == true ?
                                    /* Internal sublink */
                                    <NavLink
                                        className="MobileNavigationLink__sublink_link"
                                        to={sublink.href}
                                        onClick={
                                            this.sublink_Menu__Toggle &
                                            this.navMenu_Mobile__Toggle
                                        }
                                    >
                                        {sublink.text}
                                    </NavLink>
                                    :

                                    /* External sublink */
                                    <a
                                        className="MobileNavigationLink__sublink_link"
                                        href={sublink.href}
                                        onClick={
                                            this.sublink_Menu__Toggle &
                                            this.navMenu_Mobile__Toggle
                                        }
                                    >
                                        {sublink.text}
                                    </a>
                                }
                            </li>
                        ))}
                    </ul>
                }
            </li>
        );
    };
};


/* Connect to store */
const mapDispatchToProps = (dispatch) => ({
    navMenu_Mobile__Toggle: () => {
        dispatch(navMenu_Mobile__Toggle());
    }
});


export default connect(undefined, mapDispatchToProps)(MobileNavigationLink);