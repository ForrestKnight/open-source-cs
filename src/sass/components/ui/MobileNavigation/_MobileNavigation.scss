#MobileNavigation {
    @media #{$desktop_breakpoint} {
        display: none;
    }
}


.MobileNavigation__button {
    cursor: pointer;
    position: fixed;
    top: 10px;
    right: 12px;
    font-size: 2.5em;
    z-index: 8;

    &[data-navMenu_Mobile=false] {
        color: $primaryColor;
    }

    &[data-navMenu_Mobile=true] {
        color: $white;
    }
}


.MobileNavigation__menu {
    position: fixed;
    top: 0;
    right: 0;

    &[data-active=false] {
        opacity: 0;
        transition: 200ms;
        width: 0;
        height: 0;
    }

    &[data-active=true] {
        display: block;
        opacity: 1;
        transition: 200ms;
        background: desaturate(darken($primaryColor, 35%), 70%);
        z-index: 7;
        width: 100%;
        height: 100%;

        @media #{$tablet_breakpoint} {
            width: 350px;
        }
    }
}


.MobileNavigation__list {
    padding: 0;
    margin-top: 80px;

    &[data-active=false] {
        opacity: 0;
    }

    &[data-active=true] {
        opacity: 1;
        transition: opacity 200ms ease-in;
    }

    @media #{$tablet_breakpoint} {
        margin-top: 100px;
    }
}