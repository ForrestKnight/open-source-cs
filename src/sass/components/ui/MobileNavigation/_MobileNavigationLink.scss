.MobileNavigationLink {
    list-style-type: none;
    padding: 0;
    background: $shadow_light;
    min-height: 3em;
    
    &:not(:last-child) {
        margin-bottom: 1em;
    }
}


.MobileNavigationLink__link {
    text-decoration: none;
    color: $white;
    font-size: 1.5em;
    text-align: left;
    float: left;
    margin: 0;
    width: calc(100% - calc(3em + 3px));
    padding: 10px;
}


.MobileNavigationLink__sublink_button {
    z-index: 2;
    float: right;
    height: 3em;
    width: calc(3em + 3px);
    margin: 0;
    padding: 0;
    color: $white;
    border: 0;
    background: transparent;

    &[data-active=false] {
        border-left: 3px solid $shadow;
    }

    &[data-active=true] {
        transform: rotate(180deg);
        border-right: 3px solid $shadow;
    }

    &:active, &:focus {
        outline: 0px;
    }
}


.MobileNavigationLink__sublink_list {
    list-style-type: none;
    width: 100%;
    overflow: hidden;
    padding-left: 0;

    &[data-active=false] {
        height: 0;
        color: transparent;
        background-color: transparent;
    }

    &[data-active=true] {
        height: auto;
        background-color: $shadow;
        color: $white;
    }
}


.MobileNavigationLink__sublink {
    float: left;
    clear: both;
    padding: 0;
    width: 100%;
    
    &:not(:last-child) {
        border-bottom: 1px solid $shadow;
    }
}

.MobileNavigationLink__sublink_link {
    width: 100%;
    margin: 0;
    padding: 10px;
    padding-left: 30px;
    text-decoration: none;
    color: $white;
    font-size: 1.5em;
    text-align: left;
    float: left;
    margin: 0;
    width: 100%;
}