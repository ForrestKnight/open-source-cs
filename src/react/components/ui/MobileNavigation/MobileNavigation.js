/* Packages */
import React from 'react';
import {connect} from 'react-redux';


/* Actions */
import {navMenu_Mobile__Toggle} from '../../../redux/actions/ui';


/* Components */
import MobileNavigationLink from './MobileNavigationLink';


/* Component */
export class MobileNavigation extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            navMenu_Mobile: false
        };
    };

    
    navMenu_Mobile__Toggle = () => {
        /* Disable/enable scrolling */
        const html = document.getElementsByTagName('HTML')[0];
        if (this.props.navMenu_Mobile) {
            html.setAttribute('data-Mobile_Nav', 'false');
        } else {
            html.setAttribute('data-Mobile_Nav', 'true');
        }
        
        /* Set state */
        this.props.navMenu_Mobile__Toggle();
        this.setState({
            navMenu_Mobile: !this.state.navMenu_Mobile
        });
    };


    render() {
        return (
            <section id="MobileNavigation">
                <div 
                    className="MobileNavigation__button"
                    onClick={this.navMenu_Mobile__Toggle}
                    data-navMenu_Mobile={this.props.navMenu_Mobile}  
                >
                    {this.props.navMenu_Mobile ? "✕" : "☰"}
                </div>

                <nav 
                    data-active={this.props.navMenu_Mobile}
                    className="MobileNavigation__menu"
                >
                    <ul 
                        className="MobileNavigation__list"
                        data-active={this.props.navMenu_Mobile}
                    >
                        {/* Links */}
                        {this.props.links &&
                            this.props.links.map((link) => (
                                <MobileNavigationLink 
                                    key={link.text}
                                    {...link}
                                />
                            ))
                        }
                    </ul>
                </nav>
            </section>
        );
    };
};


/* Connect to store */
const mapStateToProps = (state) => {
    return {
        navMenu_Mobile: state.ui.navMenu_Mobile
    };
};


const mapDispatchToProps = (dispatch) => ({
    navMenu_Mobile__Toggle: () => {
        dispatch(navMenu_Mobile__Toggle());
    }
});


export default connect(mapStateToProps, mapDispatchToProps)(MobileNavigation);