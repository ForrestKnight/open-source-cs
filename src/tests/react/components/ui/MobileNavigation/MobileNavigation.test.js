/* Packages */
import React from 'react';
import {shallow} from 'enzyme';


/* Component */
import {MobileNavigation} from '../../../../../react/components/ui/MobileNavigation/MobileNavigation';


/* Test Config */
let component, navMenu_Mobile__Toggle;
beforeEach(() => {
    navMenu_Mobile__Toggle = jest.fn();
    component = shallow(
        <MobileNavigation 
            navMenu_Mobile__Toggle={navMenu_Mobile__Toggle}
        />
    );
});


/* Tests */
test('Render MobileNavigation component', () => {
    expect(component).toMatchSnapshot();
    expect(component.state('navMenu_Mobile')).toBe(false);
});


test('Toggle nav menu', () => {
    component.find('.MobileNavigation__button').simulate('click');
    expect(navMenu_Mobile__Toggle).toHaveBeenCalled();
});

