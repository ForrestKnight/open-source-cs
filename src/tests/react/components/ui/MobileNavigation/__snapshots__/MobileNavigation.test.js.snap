// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render MobileNavigation component 1`] = `
<section
  id="MobileNavigation"
>
  <div
    className="MobileNavigation__button"
    onClick={[Function]}
  >
    â˜°
  </div>
  <nav
    className="MobileNavigation__menu"
  >
    <ul
      className="MobileNavigation__list"
    />
  </nav>
</section>
`;
